### PR TITLE
Move error tolerance logic to Supervisor

### PR DESCRIPTION
**Pull request by roman**
### Context

As part of the work required to support `OneToAll` restarts, we need to move the error tolerance check logic from workers to supervisors.

Initially, worker instances would contain the state about errors registered in the supervision tree. Using this approach adds accidental complexity when dealing with restarts on siblings, as we need to aggregate error counts for each workers and make sure we are dealing with state correctly. 

By moving this logic to the supervisor, now we only need to keep track of error counts and error time windows in a single spot (the supervisor).

### CODE REVIEW

File: cap/err_tolerance.go [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+	"time"
+)
+
+// ErrToleranceResult indicates the result of a error tolerance check
+type ErrToleranceResult uint32
+
+const (
+	// ErrToleranceSurpassed indicates the error tolerance has been surpassed
+	ErrToleranceSurpassed = iota
+	// IncreaseErrCount indicates that we should allow the error to happen
```
Comment (julio-garcia-fc) [inline cap/err_tolerance.go:13]: Why not call it `AllowError`? Do you mean something like 

`IncreaseErrorCount will increase the count of errors observed`?

File: cap/err_tolerance.go [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+	ErrWindow   time.Duration
+}
+
+func (et ErrTolerance) isWithinErrorWindow(createdAt time.Time) bool {
+	// when errWindow is 0, it means we never forget errors happened
+	return time.Since(createdAt) < et.ErrWindow || et.ErrWindow == 0
+}
+
+func (et ErrTolerance) didSurpassErrorCount(restartCount uint32) bool {
+	return et.MaxErrCount < restartCount
```
Comment (julio-garcia-fc) [inline cap/err_tolerance.go:44]: Is `err` synonyms with restart? If so, why not call it `restart`? `MaxRestartCount`, `RestartWindow`, ...

File: cap/error.go [hunk 3]
```diff
@@ -167,3 +168,49 @@ func (err *SupervisorRestartError) KVs() map[string]interface{} {
 
 	return acc
 }
+
+// ErrorToleranceReached is an error that gets reported when a supervisor has
+// restarted a child so many times over a period of time that it does not make
+// sense to keep restarting.
+type ErrorToleranceReached struct {
```
Comment (julio-garcia-fc) [inline cap/error.go:175]: I would really use the terminology `RestartToleranceReached`. What if there is an evolution where errors do not result in a restart?

File: cap/one_for_one.go [hunk 4]
```diff
@@ -42,40 +38,38 @@ func oneForOneRestartLoop(
...
 	supChildren map[string]c.Child,
 	supNotifyCh chan<- c.ChildNotification,
-	wasComplete bool,
 	prevCh c.Child,
 	prevErr error,
-) *c.ErrorToleranceReached {
+) *ErrorToleranceReached {
 	for {
+		if prevErr != nil {
+			ok := supTolerance.checkTolerance()
```
Comment (julio-garcia-fc) [inline cap/one_for_one.go:49]: I can't exactly put my finger on it, but it feels like the `prevErr` should be encapsulated in the tolerance?

File: cap/one_for_one.go [hunk 5]
```diff
@@ -42,40 +38,38 @@ func oneForOneRestartLoop(
...
+	supTolerance *errToleranceManager,
 	supChildren map[string]c.Child,
 	supNotifyCh chan<- c.ChildNotification,
-	wasComplete bool,
 	prevCh c.Child,
 	prevErr error,
-) *c.ErrorToleranceReached {
+) *ErrorToleranceReached {
 	for {
+		if prevErr != nil {
```
Comment (julio-garcia-fc) [inline cap/one_for_one.go:48]: The overloading of the `prevErr` parameter and the mutation in the loop is a little convoluted. The `errToleranceManager` could keep track of the last error?

**Review by julio-garcia-fc — APPROVED** [review]
Just a few thoughts

File: cap/one_for_one.go (hunk 5 above)
Comment (roman) [reply cap/one_for_one.go:48]: Good idea, let me work with that and see how it feels

Comment (roman) [reply cap/one_for_one.go:48]: Uhm, the `prevErr` is a misnomer, I should call it `prevChErr` instead

File: cap/one_for_one.go (hunk 4 above)
Comment (roman) [reply cap/one_for_one.go:54]: Check the comment on the `prevErr` declaration above

**Review by julio-garcia-fc — APPROVED** [review]
Thanks for considering my feedback The rename really helped.

**Pull request merged**